src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -64491,6 +64491,79 @@ void M(T parameter)
             comp.VerifyDiagnostics();
         }
 
+        [Fact]
+        public void ForEach_NullableGetEnumerator_01()
+        {
+            var source = @"
+using System.Collections;
+
+class C
+{
+    public IEnumerator? GetEnumerator() => throw null!;
+
+    static void M1(C c)
+    {
+        foreach (var obj in c) // 1
+        {
+        }
+    }
+
+    static void M2(C c)
+    {
+        foreach (var obj in c!)
+        {
+        }
+    }
+}";
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (10,29): warning CS8602: Dereference of a possibly null reference.
+                //         foreach (var obj in c) // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c").WithLocation(10, 29));
+        }
+
+        [Fact]
+        public void ForEach_NullableGetEnumerator_02()
+        {
+            var source = @"
+using System.Collections;
+
+class C
+{
+    public IEnumerator? GetEnumerator() => throw null!;
+
+    static void M1(C? c)
+    {
+        foreach (var obj in c) // 1
+        {
+        }
+    }
+
+    static void M2(C? c)
+    {
+        C c2 = c!;
+        foreach (var obj in c2) // 2
+        {
+        }
+    }
+
+    static void M3(C? c)
+    {
+        foreach (var obj in c!)
+        {
+        }
+    }
+}";
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                    // (10,29): warning CS8602: Dereference of a possibly null reference.
+                    //         foreach (var obj in c) // 1
+                    Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c").WithLocation(10, 29),
+                    // (18,29): warning CS8602: Dereference of a possibly null reference.
+                    //         foreach (var obj in c2) // 2
+                    Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c2").WithLocation(18, 29));
+        }
```

Comment:
Do we have tests verifying the behavior when the generic form of enumeration?

Reply:
Looks like no.

Reply:
Consider testing an enumerable pattern as well.
```C#
class Enumerator
{
    public object Current => null!;
    public bool MoveNext() => false;
}
class Enumerable
{
    public Enumerator? GetEnumerator() => null;
}
```

---
In reply to: [322958410](https://github.com/dotnet/roslyn/pull/38613#discussion_r322958410) [](ancestors = 322958410)

Reply:
I added the test anyway, since I did wonder if a pattern-based enumerator type would make a difference.